Fix handling of deprecated timezones (Trac 56468)

jrfnl opened the pull request:
### Tests: reset timezone related options if the tests change them

The `options` table is **not** reset after each test/test class, so if an option is changed during a tests, it should be reset to the default value _after_ the test.

This commit does so for those tests which didn't have such resetting in place yet, while one or more tests in the class _do_ change the value of the `timezone_string` option.

Follow up on Trac#45821, [45857]

### Tests: replace the timezone used in tests

The `Europe/Kiev` timezone has become deprecated in PHP 8.2 and been replaced with `Europe/Kyiv`.

These tests are testing WP date/time functionality. They are **not** testing whether WP/PHP can handle deprecated timezone names correctly.

To ensure the tests stay "pure" and test what the original purpose was for these tests, I'm replacing the use of `Europe/Kiev` within these tests now with the `Europe/Helsinki` timezone, which is within the same timezone as `Europe/Kyiv`.
This should ensure that these tests run without issue and test what they are supposed to be testing on every supported PHP version (unless at some point in the future `Europe/Helsinki` would be renamed, but that's a bridge to cross if and when).

Note: separate tests should/will be added to ensure that relevant date/time related functions handle a deprecated timezone correctly, but that is not something _these_ tests are supposed to be testing.

### Tests: add tests with deprecated timezone strings

This commit adds tests in select places to safeguard that these date/time related functions will continue to behave as expected when the `timezone_string` option has been set to an outdated/deprecated timezone name.

The timezone string I'm using in these tests, `America/Buenos_Aires`, is a timezone string which was already deprecated in PHP 5.6.20 (the current minimum PHP version), so using this timezone string, we can safely test the handling of deprecated timezone names on all supported PHP versions.

See: https://3v4l.org/Holsr#v5.6.20

### I18N: Update list of continents and cities for the timezone selection

Based on a two-way comparison between the available timezone city names in PHP 5.6.20 and PHP 8.2.0.
Lists of available timezone names retrieved using the PHP `timezone_identifiers_list()` function.

See: https://3v4l.org/ro1vY/rfc#vgit.master

Note: both spellings of `Kiev`/`Kyiv` need to be in the list to allow it to work PHP cross-version.
The "old" version - `Kiev` - will be used as the basis to find the localized name for the timezone drop down lists on PHP 5.6 - 8.1, while the corrected spelling - `Kyiv` - will be used when to find the localized name for the timezone drop down lists on PHP 8.2 and up.

Previous: Trac#52861 / [50555].

### sanitize_option(): fix bug in sanitization of timezone_string

This fixes a bug where if the `timezone_string` would be set to timezone name which has since been deprecated, the option value would be "lost" when saving the value again, as the comparison being done to verify whether it is a valid timezone name would only take "current" timezone names into account and would invalidate deprecated timezone names.

By passing the `DateTimeZone::ALL_WITH_BC` constant as the `$timezoneGroup` parameter to the PHP native `timezone_identifiers_list()` function, a timezone name list is retrieved containing both current and deprecated timezone names, preventing the invalidation of the option value.

See the extensive write-up about this in Trac#56468.

Also see: https://www.php.net/manual/en/datetimezone.listidentifiers.php

Includes adding a dedicated test to the data provider used in the `Tests_Option_SanitizeOption` test class.

Note: I have made this a _named data set_, even though the other data sets are unnamed, to make sure it is clear what this data set is testing.

Adding test names for the original data sets in this data provider would be a great future improvement, but is outside of the scope of this commit/ticket.

### populate_options(): fix bug in sanitization of localized default `timezone_string`

This fixes a bug where if the default `timezone_string` would be set to a deprecated timezone name due to a localization providing an outdated timezone name string, this localized timezone string would be discarded and an empty string would be set as the timezone value instead.

By passing the `DateTimeZone::ALL_WITH_BC` constant as the `$timezoneGroup` parameter to the PHP native `timezone_identifiers_list()` function, a timezone name list is retrieved containing both current and deprecated timezone names, preventing the invalidation of the option value.

See the extensive write-up about this in Trac#56468.

Also see: https://www.php.net/manual/en/datetimezone.listidentifiers.php

Includes an expansion of the translators comment to encourage translators to use "old" names over "new" names.

Includes adding a dedicated test to the `Tests_Admin_IncludesSchema` test class.

### wp_timezone_choice(): fix bug in timezone dropdown list creation

This fixes a bug where if the `timezone_string` would be set to timezone name which has since been deprecated, no option would be (pre-)selected in the generated dropdown list and when the form using the dropdown list would be submitted, the "old", originally saved value would be lost as the form would submit without a value being selected for the `timezone_string` field.

The fix is a little hacky: it basically checks ahead of generating the actual dropdown list whether the `$selected_zone` value would be recognized and set to "selected" and if not, verifies the value _is_ a valid value, but outdated timezone name value and if so, adds an extra dropdown entry to the top of the list with the original value and sets this value to "selected".

See the extensive write-up about this in Trac#56468.

Also see: https://www.php.net/manual/en/datetimezone.listidentifiers.php

Note: there are no pre-existing tests at all for this method and adding a complete set of tests for this method is outside the scope of this ticket, so this fix does not contain any tests.

### Options/General page: minor tweak to support deprecated timezones

Underneath the time zone selector on the Options/General page, a small snippet of info about the selected time zone is displayed.

![image](https://user-images.githubusercontent.com/663378/187558126-0546375b-b092-41cb-b335-1c617b24fc2c.png)


This information would be missing if the timezone would be set to a deprecated timezone value, even though PHP is perfectly capable of generating that information, including for deprecated timezones.

By passing the `DateTimeZone::ALL_WITH_BC` constant as the `$timezoneGroup` parameter to the PHP native `timezone_identifiers_list()` function, a timezone name list is retrieved containing both current and deprecated timezone names, preventing the condition from failing when the current timezone is a deprecated one.

See the extensive write-up about this in Trac#56468.

Also see: https://www.php.net/manual/en/datetimezone.listidentifiers.php

As this is an admin/output page, no tests are available or can be set up without jumping through a lot of hoops.

Trac ticket: https://core.trac.wordpress.org/ticket/56468

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

jrfnl committed:
Tests: reset timezone related options if the tests change them

The `options` table is **not** reset after each test/test class, so if an option is changed during a tests, it should be reset to the default value _after_ the test.

This commit does so for those tests which didn't have such resetting in place yet, while one or more tests in the class _do_ change the value of the `timezone_string` option.

Follow up on Trac#45821, [45857]

jrfnl committed:
sanitize_option(): fix bug in sanitization of `timezone_string`

This fixes a bug where if the `timezone_string` would be set to timezone name which has since been deprecated, the option value would be "lost" when saving the value again, as the comparison being done to verify whether it is a valid timezone name would only take "current" timezone names into account and would invalidate deprecated timezone names.

By passing the `DateTimeZone::ALL_WITH_BC` constant as the `$timezoneGroup` parameter to the PHP native `timezone_identifiers_list()` function, a timezone name list is retrieved containing both current and deprecated timezone names, preventing the invalidation of the option value.

See the extensive write-up about this in Trac#56468.

Also see: https://www.php.net/manual/en/datetimezone.listidentifiers.php

Includes adding a dedicated test to the data provider used in the `Tests_Option_SanitizeOption` test class.

Note: I have made this a _named data set_, even though the other data sets are unnamed, to make sure it is clear what this data set is testing.

Adding test names for the original data sets in this data provider would be a great future improvement, but is outside of the scope of this commit/ticket.

jrfnl committed:
I18N: Update list of continents and cities for the timezone selection.

Based on a two-way comparison between the available timezone city names in PHP 5.6.20 and PHP 8.2.0.
Lists of available timezone names retrieved using the PHP `timezone_identifiers_list()` function.

See: https://3v4l.org/ro1vY/rfc#vgit.master

Note: both spellings of `Kiev`/`Kyiv` need to be in the list to allow it to work PHP cross-version.
The "old" version - `Kiev` - will be used as the basis to find the localized name for the timezone drop down lists on PHP 5.6 - 8.1, while the corrected spelling - `Kyiv` - will be used when to find the localized name for the timezone drop down lists on PHP 8.2 and up.

Previous: Trac#52861 / [50555].

jrfnl committed:
Tests: replace the timezone used in tests

The `Europe/Kiev` timezone has become deprecated in PHP 8.2 and been replaced with `Europe/Kyiv`.

These tests are testing WP date/time functionality. They are **not** testing whether WP/PHP can handle deprecated timezone names correctly.

To ensure the tests stay "pure" and test what the original purpose was for these tests, I'm replacing the use of `Europe/Kiev` within these tests now with the `Europe/Helsinki` timezone, which is within the same timezone as `Europe/Kyiv`.
This should ensure that these tests run without issue and test what they are supposed to be testing on every supported PHP version (unless at some point in the future `Europe/Helsinki` would be renamed, but that's a bridge to cross if and when).

Note: separate tests should/will be added to ensure that relevant date/time related functions handle a deprecated timezone correctly, but that is not something _these_ tests are supposed to be testing.

jrfnl committed:
Tests: add tests with deprecated timezone strings

This commit adds tests in select places to safeguard that these date/time related functions will continue to behave as expected when the `timezone_string` option has been set to an outdated/deprecated timezone name.

The timezone string I'm using in these tests, `America/Buenos_Aires`, is a timezone string which was already deprecated in PHP 5.6.20 (the current minimum PHP version), so using this timezone string, we can safely test the handling of deprecated timezone names on all supported PHP versions.

See: https://3v4l.org/Holsr#v5.6.20

jrfnl committed:
populate_options(): fix bug in sanitization of localized default `timezone_string`

This fixes a bug where if the default `timezone_string` would be set to a deprecated timezone name due to a localization providing an outdated timezone name string, this localized timezone string would be discarded and an empty string would be set as the timezone value instead.

By passing the `DateTimeZone::ALL_WITH_BC` constant as the `$timezoneGroup` parameter to the PHP native `timezone_identifiers_list()` function, a timezone name list is retrieved containing both current and deprecated timezone names, preventing the invalidation of the option value.

See the extensive write-up about this in Trac#56468.

Also see: https://www.php.net/manual/en/datetimezone.listidentifiers.php

Includes an expansion of the translators comment to encourage translators to use "old" names over "new" names.

Includes adding a dedicated test to the `Tests_Admin_IncludesSchema` test class.

jrfnl committed:
wp_timezone_choice(): fix bug in timezone dropdown list creation

This fixes a bug where if the `timezone_string` would be set to timezone name which has since been deprecated, no option would be (pre-)selected in the generated dropdown list and when the form using the dropdown list would be submitted, the "old", originally saved value would be lost as the form would submit without a value being selected for the `timezone_string` field.

The fix is a little hacky: it basically checks ahead of generating the actual dropdown list whether the `$selected_zone` value would be recognized and set to "selected" and if not, verifies the value _is_ a valid value, but outdated timezone name value and if so, adds an extra dropdown entry to the top of the list with the original value and sets this value to "selected".

See the extensive write-up about this in Trac#56468.

Also see: https://www.php.net/manual/en/datetimezone.listidentifiers.php

Note: there are no pre-existing tests at all for this method and adding a complete set of tests for this method is outside the scope of this ticket, so this fix does not contain any tests.

jrfnl committed:
Options/General page: minor tweak to support deprecated timezones

Underneath the time zone selector on the Options/General page, a small snippet of info about the selected time zone is displayed.

This information would be missing if the timezone would be set to a deprecated timezone value, even though PHP is perfectly capable of generating that information, including for deprecated timezones.

By passing the `DateTimeZone::ALL_WITH_BC` constant as the `$timezoneGroup` parameter to the PHP native `timezone_identifiers_list()` function, a timezone name list is retrieved containing both current and deprecated timezone names, preventing the condition from failing when the current timezone is a deprecated one.

See the extensive write-up about this in Trac#56468.

Also see: https://www.php.net/manual/en/datetimezone.listidentifiers.php

As this is an admin/output page, no tests are available or can be set up without jumping through a lot of hoops.

costdev commented on tests/phpunit/tests/admin/includesSchema.php:
```diff
@@ -177,6 +177,47 @@ public function data_populate_options() {
+		// Ensure the "default" value for the timezone will be set to a deprecated timezone.
+		add_filter(
+			'gettext_with_context',
+			function( $translation, $text, $context ) {
```
```suggestion
			static function( $translation, $text, $context ) {
```

costdev commented on tests/phpunit/tests/date/currentTime.php:
```diff
@@ -117,4 +128,33 @@ public function test_should_return_correct_local_time() {
+		$current_time_gmt = current_time( $format, true );
+		$current_time     = current_time( $format );
+
+		$this->assertSame( gmdate( $format ), $current_time_custom_timezone_gmt, 2, 'The dates should be equal [1]' );
+		$this->assertSame( $datetime->format( $format ), $current_time_custom_timezone, 2, 'The dates should be equal [2]' );
+		$this->assertSame( gmdate( $format ), $current_time_gmt, 2, 'The dates should be equal [3]' );
+		$this->assertSame( $datetime->format( $format ), $current_time, 2, 'The dates should be equal [4]' );
```
For these assertions, the `message` value is `2`, rather than the provided string. Not sure if I'm missing something here though.

jrfnl replied:
Good catch! I transformed those from similar assertions which were using `assertSameWithDelta()` and clearly overlooked removing the "delta". Serves me right for working silly hours.

costdev commented on tests/phpunit/tests/date/getFeedBuildDate.php:
```diff
@@ -11,7 +11,7 @@ class Tests_Date_GetFeedBuildDate extends WP_UnitTestCase {
 		global $wp_query;
 
-		update_option( 'timezone_string', 'UTC' );
+		update_option( 'timezone_string', '' );
```
In the `Test_Date_GetPermalink` class, [the `tear_down()`](https://github.com/WordPress/wordpress-develop/blob/a2a63479d1f18d6108408d02a6ba983d4300853a/tests/phpunit/tests/date/getPermalink.php#L14-L15) calls `date_default_timezone_set( 'UTC' );` afterwards. Is this something that should be added here also?

jrfnl replied:
No. The test function in getPermalink calls the `date_default_timezone_set()` function, so resetting it, is prudent, but the tests in this file, don't call that function.

costdev commented on tests/phpunit/tests/date/getPostTime.php:
```diff
@@ -9,6 +9,16 @@
+	 */
+	public function tear_down() {
+		// Reset the timezone option to the default value.
+		update_option( 'timezone_string', '' );
```
Same question as above regarding `date_default_timezone_set( 'UTC' );`.

jrfnl replied:
same answer

costdev commented on tests/phpunit/tests/date/mysql2date.php:
```diff
@@ -11,6 +11,9 @@ public function tear_down() {
 		date_default_timezone_set( 'UTC' );
 
+		// Reset the timezone option to the default value.
+		update_option( 'timezone_string', '' );
```
Slightly different one this time, `date_default_timezone_set( 'UTC' );` is called before `update_option()`. Is this worth changing for consistency, or not?

+ If both of these should always be used, would a new helper (called something like `::reset_timezone()`) be useful in the test suite? If so, not sure if that's in scope for this or better handled in a separate ticket.

jrfnl replied:
Outside of the scope of this ticket. (And most consistency between test classes is incidental, not by design)

costdev commented on tests/phpunit/tests/date/mysql2date.php:
```diff
@@ -59,11 +62,27 @@ public function test_mysql2date_should_format_time_with_changed_time_zone() {
+		$rfc3339  = $datetime->format( DATE_RFC3339 );
+		$mysql    = $datetime->format( 'Y-m-d H:i:s' );
+
+		$this->assertSame( $rfc3339, mysql2date( DATE_RFC3339, $mysql ) );
+		$this->assertSame( $rfc3339, mysql2date( DATE_RFC3339, $mysql, false ) );
```
These assertions should have a `message` parameter, or the test should use a data provider to test when `$translate` is `true` and `false`.

jrfnl replied:
Good point, but one which should be actioned for the whole test file (all tests are missing the `$message` param, so best to fix them all in one go with consistent messages, which would be out of scope for this ticket).

costdev commented on tests/phpunit/tests/date/query.php:
```diff
@@ -21,6 +21,16 @@ public function set_up() {
+	 */
+	public function tear_down() {
+		// Reset the timezone option to the default value.
+		update_option( 'timezone_string', '' );
```
Same question as earlier regarding `date_default_timezone_set( 'UTC' );`.

jrfnl replied:
same answer

costdev commented on tests/phpunit/tests/date/wpTimezone.php:
```diff
@@ -8,6 +8,17 @@
+	public function tear_down() {
+		// Reset changed options to their default value.
+		update_option( 'gmt_offset', 0 );
+		update_option( 'timezone_string', '' );
```
Same question as above regarding `date_default_timezone_set( 'UTC' );`.

jrfnl replied:
same answer

costdev commented on tests/phpunit/tests/date/wpTimezone.php:
```diff
@@ -31,13 +42,29 @@ public function test_should_convert_gmt_offset( $gmt_offset, $tz_name ) {
-		update_option( 'timezone_string', 'Europe/Kiev' );
+		update_option( 'timezone_string', 'Europe/Helsinki' );
+
+		$this->assertSame( 'Europe/Helsinki', wp_timezone_string() );
```
These assertions should have a `$message` parameter.

jrfnl replied:
Agreed, but improving the tests as-such is outside the scope of this ticket. This commit only changes the timezone name used, nothing else.

costdev commented on tests/phpunit/tests/date/wpTimezone.php:
```diff
@@ -31,13 +42,29 @@ public function test_should_convert_gmt_offset( $gmt_offset, $tz_name ) {
+		update_option( 'timezone_string', $tz_string );
 
-		$this->assertSame( 'Europe/Kiev', wp_timezone_string() );
+		$this->assertSame( $tz_string, wp_timezone_string() );
```
These assertions should have a `$message` parameter.

jrfnl replied:
https://github.com/WordPress/wordpress-develop/pull/3155#discussion_r960406265

costdev commented on tests/phpunit/tests/date/xmlrpc.php:
```diff
@@ -8,6 +8,16 @@
+	 */
+	public function tear_down() {
+		// Reset the timezone option to the default value.
+		update_option( 'timezone_string', '' );
```
Same question as earlier regarding `date_default_timezone_set( 'UTC' );`.

jrfnl replied:
same answer

costdev commented on tests/phpunit/tests/formatting/date.php:
```diff
@@ -6,6 +6,17 @@
+	public function tear_down() {
+		// Reset changed options to their default value.
+		update_option( 'gmt_offset', 0 );
+		update_option( 'timezone_string', '' );
```
Same question as earlier regarding `date_default_timezone_set( 'UTC' );`.

jrfnl replied:
same answer

costdev commented on tests/phpunit/tests/option/sanitizeOption.php:
```diff
@@ -69,6 +69,7 @@ public function sanitize_option_provider() {
 			array( 'timezone_string', 0, 0 ),
 			array( 'timezone_string', 'Europe/London', 'Europe/London' ),
 			array( 'timezone_string', get_option( 'timezone_string' ), 'invalid' ),
+			'Deprecated timezone string is accepted as valid (Trac#56468)' => array( 'timezone_string', 'America/Buenos_Aires', 'America/Buenos_Aires' ),
```
😆 When you can't bring yourself to introduce a dataset without a key. I feel that.

costdev commented on tests/phpunit/tests/date/currentTime.php:
```diff
@@ -7,6 +7,17 @@
 class Tests_Date_CurrentTime extends WP_UnitTestCase {
 
+	/**
+	 * Clean up.
```
```suggestion
	 * Cleans up.
```

costdev commented on tests/phpunit/tests/date/dateI18n.php:
```diff
@@ -7,11 +7,22 @@
 class Tests_Date_DateI18n extends WP_UnitTestCase {
 
+	/**
+	 * Clean up.
```
```suggestion
	 * Cleans up.
```

costdev commented on tests/phpunit/tests/date/getPostTime.php:
```diff
@@ -9,6 +9,16 @@
 class Tests_Date_GetPostTime extends WP_UnitTestCase {
 
+	/**
+	 * Clean up.
```
```suggestion
	 * Cleans up.
```

costdev commented on tests/phpunit/tests/date/query.php:
```diff
@@ -21,6 +21,16 @@ public function set_up() {
 	}
 
+	/**
+	 * Clean up.
```
```suggestion
	 * Cleans up.
```

costdev commented on tests/phpunit/tests/date/wpTimezone.php:
```diff
@@ -8,6 +8,17 @@
 class Tests_Date_wpTimezone extends WP_UnitTestCase {
 
+	/**
+	 * Clean up.
```
```suggestion
	 * Cleans up.
```

costdev commented on tests/phpunit/tests/date/xmlrpc.php:
```diff
@@ -8,6 +8,16 @@
 class Tests_Date_XMLRPC extends WP_XMLRPC_UnitTestCase {
 
+	/**
+	 * Clean up.
```
```suggestion
	 * Cleans up.
```

costdev commented on tests/phpunit/tests/formatting/date.php:
```diff
@@ -6,6 +6,17 @@
 class Tests_Formatting_Date extends WP_UnitTestCase {
 
+	/**
+	 * Clean up.
```
```suggestion
	 * Cleans up.
```

costdev commented on tests/phpunit/tests/formatting/date.php:
```diff
@@ -222,13 +233,18 @@ public function test_is8601_to_datetime_correct_time( $timezone_string, $gmt_off
 				'gmt_offset'      => 3,
 			),
+			// @ticket 56468.
+			'Deprecated timezone string and no gmt offset set' => array(
```
```suggestion
			'Deprecated timezone string and no GMT offset set' => array(
```
Also, genuine question about something trivial:

Should datasets begin with uppercase, or lowercase?

```
is8601 to datetime correct time with Deprecated timezone string and no GMT offset set
versus
is8601 to datetime correct time with deprecated timezone string and no GMT offset set
```

jrfnl replied:
There is currently no standard set for this. I normally would use sentence case, though it depends on the codebase.

costdev commented on tests/phpunit/tests/admin/includesSchema.php:
```diff
@@ -177,6 +177,47 @@ public function data_populate_options() {
+
+		wp_cache_delete( 'alloptions', 'options' );
+
+		$result = get_option( 'timezone_string' );
```
Is there a scenario in which `timezone_string` might _already_ be set to `America/Buenos_Aires` due a faulty test earlier in the class/test suite?

jrfnl replied:
Not that I can think off with the resets I added.

costdev commented on tests/phpunit/tests/date/dateI18n.php:
```diff
@@ -7,11 +7,22 @@
+	public function tear_down() {
+		// Reset changed options to their default value.
+		update_option( 'gmt_offset', 0 );
+		update_option( 'timezone_string', '' );
```
Oh, missed this one. Same question as earlier regarding `date_default_timezone_set( 'UTC' );`.

jrfnl replied:
same answer

costdev reviewed (changes requested):
Most of these comments are regarding the same issue. I've just marked each instance to make it easier to locate each change if it's necessary. If the change isn't necessary, then I apologise for the completely unnecessary `resolve conversation` clicks I've put you through! 😂 

The source changes look good at a glance, but I'll try to get the PR tested and will get back to you if I have any queries/concerns.

jrfnl commented:
> Most of these comments are regarding the same issue. I've just marked each instance to make it easier to locate each change if it's necessary. If the change isn't necessary, then I apologise for the completely unnecessary `resolve conversation` clicks I've put you through! 😂

Thanks for reviewing! I've fixed up those commits which needed it. Left a reply for everything else (and for some which I fixed up as well).

> The source changes look good at a glance, but I'll try to get the PR tested and will get back to you if I have any queries/concerns.

You're a ⭐ !

PrinceSagor reviewed (approved):
ক্রাইম এর বিরুদ্ধে আমার এই যুদ্ধ আমার পাশে থাকবেন স্বয়ং আল্লাহ তাআলা সত্যের জয় হবে ইনশাআল্লাহ সত্য কে দাবিয়ে রাখতে পারবেন কিন্তু একদিন আল্লাহতাআলা তা প্রকাশ করে দিবেন 

রেপ করলে ওপেন ফাঁসি এরকম ধরনের একটি আইন চালু হলে কেউ রেপ করার আগে দশবার ভাববে 

আজকে আমাদের আইন দুর্বল বলে 
মানুষ এরকম ধরনের ঘৃণ্য কাজ করার সাহস পায় 
সঠিক শক্তিশালী আইন হলে সমাজ থেকে অনেক বেশি ক্রাইম কমে যাবে

 কিন্তু সঠিক আইন করলে শয়তানদের সাহস কমে যাবে

কিন্তু কে সঠিক আইন করবে সবাইতো টাকার জন্য সমাজকে নষ্ট করে দিতে চায়  

সবাই একসাথে মিলে সঠিক আইন চালু করে দিলে মুখোশধারী শয়তান দের রাজত্ব ভেঙেচুরে যাবে 

এখনকার দুনিয়ায় সঠিক বিচার হয় কম কারণ টাকা দিয়ে বিচারের রায় বদলে ফেলা যায় 

কেউ সঠিকভাবে প্রতিবাদ করতে চায় না কারণ তাদের কি কোনো মানুষ বিপদে পড়লে অনেক মানুষই তাদের থেকে  পিছিয়ে যায় 

সবাই মিলে প্রতিবাদ করলে সমাজে ক্রাইম হবে বলে আমার মনে হয় না কিন্তু কেএ করবে সবাই তো টাকা পেলে নিজের চরিত্রটা কেউ বিক্রি করে দিবে 
  
অন্যায়ের বিরুদ্ধে হাতে হাত রেখে প্রতিবাদ করলে কেউ প্রাইম করার আগে তার মাথায় এটা কাজ করবে যে এখন ক্রাইম করলে সবাই একসাথে প্রতিবাদ করবে 

সমাজে কেউ ক্রাইম করলে সবাই একসাথে প্রতিবাদ করা উচিত বলে আমি মনে করি 

কেউ কারো বিপদে পাশে থাকে না তার জন্যই অনেক সময় অপরাধীরা ক্রাইম করেও সঠিক সাজা ভোগ করে না তার জন্য তাদের সাহস আরো বেড়ে যায় 

আমি যদি কিছু ভুল বলে থাকি তাহলে আমাকে ক্ষমা করে দিবেন 

হে আল্লাহ তাআলা সবাইকে সঠিক হেদায়েত দান করুন  

ইসলামিক আইন দিয়ে দুনিয়াটাকে সাজিয়ে নেওয়ার তৌফিক দান করুন 

ইসলাম অনেক কঠিন ধর্ম কিন্তু এই পথে অনেক শান্তি লুকিয়ে আছে এই পথেই আমাদের জান্নাত

 কিন্তু শয়তানের কাজে লিপ্ত হয়ে পড়েছি তার জন্যই আমাদের প্রতি আল্লাহতালা এত আজাব-গজব দিচ্ছেন পৃথিবীতে 

মানুষেরা আসল শান্তি কিতা ভুলে গেছে এখন মানুষ মনে করে টাকা হলেই শান্তি কিন্তু একবার ভেবে দেখুন টাকা বারলেই তার সাথে অশান্তি ও বেড়ে যায় 

কারণ টাকা ইনকাম করতে হলেই নিজের সময়টাকে অন্যের কাছে বিক্রি করে দিতে হয় 

যে মালিক আমাদের সৃষ্টি করেছেন তার জন্য সময় বের করতে গেলে কিছু খারাপ মানুষদের হাতে অনেক কথা শুনতে হয় 

কারণ আমাদের মাথাটা তো আমরা বিক্রি করে দিয়েছি টাকার কাছে তার জন্যই কিছু অসৎ মানুষ রাজত্ব করে চলেছে আমাদের পৃথিবীতে 

সবাই এক হয়ে প্রতিবাদ করলে শয়তানের রাজত্ব গুলোকে আমরা ভেঙ্গে দিতে পারব 
কিন্তু তার জন্য সবাইকে একত্র হতে হবে কতদিন অন্যায়-অত্যাচার সইবো আমরা পৃথিবীতে কখনো কি আমাদের চোখ খুলবে আমরা কি সঠিক বিবেক দিয়ে বিবেচনা করতে পারব 

কেউ যদি করতে পারে তাহলে কি আমরা সকলেই একসাথে হাতে হাত রেখে প্রতিবাদ গড়ে তুলব নাকি অন্যায়ের বিরুদ্ধে মাথা নিচু করবো 

শয়তানি কি আমাদের উপরে রাজত্ব করবে আর আমরা কি মুখ বুঝে তা চুপ করে সইবো 

আমিতো পারিনা তাহলে আপনারা কি করে সইতে পারেন আমাদের বিবেক টাকে লাড়া দিতে হবে 

আজকাল টাকা হলে তো মানুষ বিবেচনা করতে ভুলে যায় এই কারণেই আমরা সঠিক তাকে চোখে দেখেও না দেখার ভান করে চলে যাই

টাকার রাজত্বকে ভেঙ্গে   আসুন আমরা সকলে বিবেক দিয়ে বিবেচনা করতে শুরু করে দেই

তাহলে সবকিছু ঠিক হয়ে যাবে সমাজে সুখ শান্তি ফিরে আসবে বলে আমার মনে হয়  

হাতে হাত রাখুন পতিটা জীবনকে সুরক্ষিত রাখুন আজকের সমাজে সঠিক শিক্ষা নেই যেটা আছে সেটাও হলো টাকার রাজত্ব কারণ কিছু অসৎ মানুষই সমাজের রাজ করে যায় 

আর মানুষ তা গাধার মত মুখ বুজে সহ্য করে যায় কারণ তারা সঠিক বিবেচনা করতে পারেনা আমি তো একা সব ঠিক করতে পারব না আমার সাথে হাতে হাত রেখে প্রতিবাদ করতে হবে 

তাহলে আমি সব ঠিকঠাকভাবে বিবেচনা করে তুলতে পারব তখন আর শয়তান রাজত্ব করতে পারবে না টাকাই হলো অসৎ মানুষের শক্তি 

 এইটাকে একবার ভেঙ্গে দিতে পারলেই মানুষেরা তাদের সময়টাকে সঠিক কাজে লাগাতে পারবে তাই বলব বেশি বেশি নামাজ আদায় করুন মানুষকে নামাজের পথে দাওয়াত দিন 

আর সময় পেলে আল্লাহর কাছে একটি দোয়া  বেশি বেশি  করবেন তাহলো জেরও যালিমদের হাত থেকে রক্ষা করুন সঠিক শিক্ষা দান করুন আপনি ছাড়া আমাদের কেউ রক্ষা করতে পারবে না 

আমরা অনেক বেশি গুনা করে ফেলেছি তাই আজকে আপনি আমাদের প্রতি নারাজ  আমরা তো আপনার বান্দা আপনি হলেন আমাদের রব আপনি ক্ষমা না করলে কে আমাদের ক্ষমা করবে 

আমরা অনেক বেশি গুনা করে ফেলেছি যে কারণে পৃথিবীতে এত আজাব গজব চলে এসেছে তার থেকে আমাদের রক্ষা করুন 

সঠিক পথে চলার তৌফিক দান করুন এবং আপনার ইবাদতে লিপ্ত রাখুন এবং আপনার কুদরতের শক্তি দিয়ে শয়তানি শক্তি ধ্বংস করে দিন

SergeyBiryukov commented:
Thanks for the PR! Merged in [r54207](https://core.trac.wordpress.org/changeset/54207), [r54217](https://core.trac.wordpress.org/changeset/54217), [r54227](https://core.trac.wordpress.org/changeset/54227), [r54229](https://core.trac.wordpress.org/changeset/54229), [r54230](https://core.trac.wordpress.org/changeset/54230), [r54232](https://core.trac.wordpress.org/changeset/54232), [r54233](https://core.trac.wordpress.org/changeset/54233), and [r54237](https://core.trac.wordpress.org/changeset/54237).

jrfnl commented:
Thanks for getting those all in @SergeyBiryukov !
